frontend/src/app/chart/chart.component.ts

```diff
@@ -0,0 +1,255 @@
+      .area()
+      .defined((d) => +d[1] >= 0)
+      .x((d: any) => x_scale(this.time_parse(d[0])))
+      .y0(() => y_scale.range()[0])
+      .y1((d: any) => y_scale(+d[1]));
+
+    chart
+      .append('path')
+      .datum(records)
+      .classed('area', true)
+      .attr('fill', '#F2EEB3')
+      .attr('d', area);
+
+    const bisect = d3.bisector((d) => this.time_parse(d[0])).left;
+
+    // Create the circle that travels along the curve of chart
+    const focus = chart
+      .append('g')
+      .append('circle')
+      .style('fill', 'none')
+      .attr('stroke', 'black')
+      .attr('r', 8.5)
+      .style('opacity', 0);
+
+    // Create the text that travels along the curve of chart
+    const focusText = chart
+      .append('g')
+      .append('text')
+      .attr('opacity', 1)
+      .attr('text-anchor', 'left')
+      .attr('alignment-baseline', 'middle')
+      .attr('font-size', '20px');
+
+    // Draw lines
+
+    chart
+      .append('path')
+      .datum(records)
+      .classed('line', true)
+      .attr('fill', 'none')
+      .attr('stroke', '#F29F05')
+      .attr('stroke-width', 2)
+      .attr(
+        'd',
+        d3
+          .line()
+          .x((d: any) => x_scale(this.time_parse(d[0])))
+          .y((d: any) => y_scale(+d[1]))
+      )
+      .attr('opacity', 0.4);
+
+    // Brush functionality
+    const brush = d3
+      .brushX()
+      .extent([
+        [this.chart_padding, this.chart_padding],
+        [
+          this.chart_width - this.chart_padding,
+          this.chart_height - this.chart_padding,
+        ],
+      ])
+      .on('end', () => {
+        console.log('update chart');
+        updateChart();
+      });
+
+    chart
+      .append('g')
+      .attr('class', 'brush')
+      .call(brush)
+      .on('mouseover', mouseover) //this mouse over value display functionality
+      .on('mousemove', mousemove)
+      .on('mouseout', mouseout);
+
+    // What happens when the mouse move -> show the annotations at the right positions.
+    function mouseover() {
+      focus.style('opacity', 1);
+      focusText.style('opacity', 1);
+    }
+
+    function mousemove() {
+      // recover coordinate we need
+      var x0 = x_scale.invert(d3.mouse(this)[0]);
+      var i = bisect(records, x0, 1);
+      const selectedData = records[i];
+
+      focus
+        .attr('cx', x_scale(time_parse(selectedData[0])))
+        .attr('cy', y_scale(+selectedData[1]));
+      focusText
+        .html(
+          `x: ${time_format(time_parse(selectedData[0]))} - y: ${
+            selectedData[1]
+          }`
+        )
+        .attr('x', x_scale(time_parse(selectedData[0])) + 15)
+        .attr('y', y_scale(+selectedData[1]));
+    }
+    function mouseout() {
+      focus.style('opacity', 0);
+      focusText.style('opacity', 0);
+    }
+
+    // A function that set idleTimeOut to null
+    var idleTimeout;
+    function idled() {
+      idleTimeout = null;
+    }
+
+    function updateChart() {
```

Comment:
Is there a specific reason to define the updateChart method inside createChart method? Maybe make updateChart becomes a property of the chart object itself makes more sense right? Let's say in another file if you wish to call the updateChart method specifically, in your code you can never call it right?

Reply:
I agree it's a bit wierd to define update function inside create function. Here the point is that, update functions references a few variables defined in the create function, such as x-scale, y-scale, x-axis. 
I could declare a few class members to hold them, but these d3 object don't seem to be a natural fit to typescript, since for exmample, the type for x-axis is d3.Selection<SVGGElement, unknown, HTMLElement, any>. 
With my limited experience in Typescript, I'm not sure if it's normal to have type like this defined in function parameters and class members. Any advices are apreciated.

Reply:
I see. Yes D3 naturally lacks type support for TypeScript. I will look at some code in our codebase to see how other people handle it. But ideally variables like x_axis, x_scale, ...etc and methods like updateChart should be attached to one chart object/class instead of the createChart method of the chart class right? I will still recommend you to move them outside the createChart method.
Also be careful of using '-' in naming your variables, I know that d3 name variables using the same way. But in google3 those variables won't pass the lint checker and we will have to change them in the future if we would like to merge your code into our codebase.

Reply:
Sure, I already changed all the variable names to camel cases.

Reply:
Moved update function to class member now